Assets/HoloToolkit/UX/Scripts/Buttons/Button.cs

```diff
@@ -0,0 +1,488 @@
+        {
+            if (OnButtonHeld != null)
+            {
+                OnButtonHeld(gameObject);
+            }
+        }
+
+        /// <summary>
+        /// Called when something interrupts the button pressed state.
+        /// </summary>
+        protected void DoButtonCancelled()
+        {
+            if (OnButtonCancelled != null)
+            {
+                OnButtonCancelled(gameObject);
+            }
+        }
+
+        /// <summary>
+        /// Use LateUpdate to check for whether or not the hand is up
+        /// </summary>
+        public void LateUpdate()
+        {
+            if (!m_disabled && m_bLastHandVisible != m_bHandVisible)
+            {
+                OnHandVisibleChange(m_bHandVisible);
+            }
+        }
+
+        /// <summary>
+        /// Event to fire off when hand visibity changes
+        /// </summary>
+        /// <param name="visible"></param>
+        public virtual void OnHandVisibleChange(bool visible)
+        {
+            m_bLastHandVisible = visible;
+
+            ButtonStateEnum newState = ButtonState;
+
+            switch (ButtonState)
+            {
+                case ButtonStateEnum.Interactive:
+                {
+                    newState = visible ? ButtonStateEnum.Interactive : ButtonStateEnum.Observation;
+                    break;
+                }
+
+                case ButtonStateEnum.Targeted:
+                {
+                    newState = visible ? ButtonStateEnum.Targeted : ButtonStateEnum.ObservationTargeted;
+                    break;
+                }
+
+                case ButtonStateEnum.Observation:
+                {
+                    newState = visible ? ButtonStateEnum.Interactive : ButtonStateEnum.Observation;
+                    break;
+                }
+
+                case ButtonStateEnum.ObservationTargeted:
+                {
+                    newState = visible ? ButtonStateEnum.Targeted : ButtonStateEnum.ObservationTargeted;
+                    break;
+                }
+
+            }
+
+            OnStateChange(newState);
+        }
+
+        /// <summary>
+        /// Ensures the button returns to a neutral state when disabled
+        /// </summary>
+        public virtual void OnDisable()
+        {
+            if (ButtonState != ButtonStateEnum.Disabled)
+            {
+                OnStateChange(ButtonStateEnum.Observation);
+            }
+        }
+
+        /// <summary>
+        /// Callback virtual function for when the button state changes
+        /// </summary>
+        /// <param name="newState">
+        /// A <see cref="ButtonStateEnum"/> for the new button state.
+        /// </param>
+        public virtual void OnStateChange(ButtonStateEnum newState)
+        {
+            ButtonState = newState;
+
+            // Send out the action/event for the statechange
+            if (this.StateChange != null)
+                this.StateChange(newState);
+        }
+
+#if UNITY_EDITOR
```

Comment:
nit: we don't need `UNITY_EDITOR` because `OnDrawGizmos` API only gets called in editor.

Reply:
We used to get compile errors without the ifdef for editor.  I'll remove them if that's no longer the case.

Reply:
Compile errors might be from `UnityEditor` dependency.
But yeah [in the docs](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnDrawGizmos.html) it says:
>This function does not get called if the component is collapsed in the inspector